File: old-src/features/selectors/selector.ts
import chalk from 'chalk';
import ora from 'ora';
import { createQuestion, selectFromList, toggleQuestion } from '../../util/questionsFunc.js';
import { error, warn } from '../../util/symbols.js';

let addSelectors: string = '';
let selector: string = '';

export async function addSelectorsQuestion(): Promise<string> {
  let targetType: string = '';
  while (true) {
    // Choose Target
    const target = [
      '@p - Near Player',
      '@a - All Player',
      '@s - Myself',
      '@r - Random Player',
      "@n - A Nearest Player (1.21+, same '@p[sort=nearest, limit=1]')",
      'PlayerName',
    ];
    targetType = (await selectFromList('Select a target selector type:', target)).split(' ')[0];

    if (targetType === 'PlayerName') {
      while (true) {
        const playerName = await createQuestion(
          chalk.cyan('Enter the player name. Type "back" to go back.\n') +
            chalk.italic.gray(
              'Note: When it is entered, the existence of the player name will be checked in Mojang API.\nIf you skip this check, please add "!" at the beginning of the player name.\n'
            )
        );
        if (!playerName.trim()) {
          console.log(error, chalk.red('Please enter a player name.'));
          continue;
        }
        if (playerName.trim().toLowerCase() === 'back') {
          console.log(warn, chalk.yellow(' Cancelled. Back to components selection.'));
          console.log('\n');
          targetType = 'backed';
          break;
        }
        targetType = playerName.trim();

        if (!targetType.startsWith('!')) {
          // ===== 追加: Mojang API で存在確認 =====
          const spinner = ora('Checking player existence...').start();
          try {
            const response = await fetch(
              `https://api.mojang.com/users/profiles/minecraft/${targetType}`
            );
            if (response.status === 404 || response.status === 204) {
              spinner.warn(
                chalk.bgYellow.black(' WARN ') +
                  chalk.yellow(` Player "${targetType}" does not exist in Mojang database.`)
              );
              const proceed = await toggleQuestion(chalk.cyan('Use this name anyway?'));
              if (!proceed) {
                continue; // もう一度入力をやり直す
              }
            } else if (response.ok) {
              spinner.succeed(
                chalk.green(
                  `Player "${targetType}" is existed in Mojang database! Using this name.`
                )
              );
            } else {
              spinner.stop(); // レート制限(429)などの場合はスルー
            }

            // eslint-disable-next-line @typescript-eslint/no-unused-vars
          } catch (_e) {
            spinner.stop(); // ネットワークエラー等もとりあえずスルー
          }
          // =======================================
        }
        if (targetType.startsWith('!')) {
          targetType = targetType.slice(1); // 先頭の'!'を削除してプレイヤー名として使用
        }

        break;
      }
    }

    if (targetType === 'backed') {
      continue;
    }

    break;
  }
  console.log(chalk.blue(`Target:`), `${chalk.green(`${chalk.bold(targetType)}`)}`);
  console.log('\n');

  if (!targetType.startsWith('@')) {
    console.log(
      `${chalk.blue('Additional target selectors:')} ${chalk.green(`${chalk.bold('Skipped (Specific Player Name)')}`)}`
    );
    console.log('\n');
    return targetType;
  }

  // Q2.5: Ask to refine target selector
  const refineSelector = await toggleQuestion(
    chalk.cyan(
      'Do you want to add more target selectors? (e.g., distance, score, tag, team, etc.)'
    ),
    'Yes',
    'No'
  );

  const shouldRefine = refineSelector === true;

  if (shouldRefine) {
    console.log(
      `${chalk.blue('Additional target selectors:')} ${chalk.green(`${chalk.bold('Yes')}`)}`
    );

    const addedSelectors: string[] = [];
    const allSelectorTypes = [
      'distance',
      'score',
      'tag',
      'team',
      'limit',
      'level',
      'gamemode',
      'advancements',
      'predicate',
      'sort',
    ];

    let continueAdding = true;

    while (continueAdding) {
      const availableSelectors = allSelectorTypes.filter(
        (selector) => !addedSelectors.some((added) => added.split('=')[0] === selector)
      );

      const selectorOptions = [
        ...availableSelectors.map((s) => {
          const descriptions: { [key: string]: string } = {
            distance: 'Distance to Entity(=Player)',
            score: 'The score value or range which the entity has',
            tag: 'The tag which the entity has',
            team: 'The team which the entity joins',
            limit: 'Amount limit',
            level: 'Experience level',
            gamemode: 'Player gamemode',
            advancements: 'The advancements which the player has',
            predicate: 'Match predicates(required datapacks)',
            sort: 'Specify the order in which to select targets',
          };
          return `${s} - ${descriptions[s] || s}`;
        }),
        'OK',
      ];

      // Q2.55 Selector Type
      const addSelectorsType = (
        await selectFromList('Additional selectors: ', selectorOptions)
      ).split(' ')[0];

      console.log(
        chalk.blue(`Additional selector:`),
        `${chalk.green(`${chalk.bold(addSelectorsType)}`)}`
      );
      console.log('\n');

      switch (addSelectorsType) {
        case 'distance': {
          while (true) {
            const distance = await createQuestion(
              chalk.cyan(
                'Distance to entity (int or range e.g., 1..5 = 1~5). Type "back" to go back: '
              )
            );
            if (distance.trim().toLowerCase() === 'back') {
              console.log(chalk.yellow('Cancelled. Back to selector selection.'));
              console.log('\n');
              break;
            }
            if (!distance.trim()) {
              console.log(chalk.red('Please enter a distance value or range.'));
              continue;
            }
            console.log(chalk.blue(`Distance:`), `${chalk.green(`${chalk.bold(distance)}`)}`);
            console.log('\n');
            addedSelectors.push(`distance=${distance}`);
            break;
          }
          break;
        }
        case 'score': {
          while (true) {
            const score = await createQuestion(
              chalk.cyan(
                "Score (format: score value or range of A = 'A=1' or 'A=1..10'). Type \"back\" to go back: "
              )
            );
            if (score.trim().toLowerCase() === 'back') {
              console.log(chalk.yellow('Cancelled. Back to selector selection.'));
              console.log('\n');
              break;
            }
            if (!score.trim()) {
              console.log(chalk.red('Please enter a score format.'));
              continue;
            }
            console.log(chalk.blue(`Score:`), `${chalk.green(`${chalk.bold(score)}`)}`);
            console.log('\n');
            addedSelectors.push(`score=${score}`);
            break;
          }
          break;
        }
        case 'tag': {
          while (true) {
            const tag = await createQuestion(
              chalk.cyan(
                'Tag (format: <your-tag> | If it put \'!\' at the beginning, the tag will be excluded.). Type "back" to go back: '
              )
            );
            if (tag.trim().toLowerCase() === 'back') {
              console.log(chalk.yellow('Cancelled. Back to selector selection.'));
              console.log('\n');
              break;
            }
            if (!tag.trim()) {
              console.log(chalk.red('Please enter a tag.'));
              continue;
            }
            console.log(chalk.blue(`Tag:`), `${chalk.green(`${chalk.bold(tag)}`)}`);
            console.log('\n');
            addedSelectors.push(`tag=${tag}`);
            break;
          }
          break;
        }
        case 'team': {
          while (true) {
            const team = await createQuestion(
              chalk.cyan(
                'Team (format: <your-team> | If it put \'!\' at the beginning, the team will be excluded.). Type "back" to go back: '
              )
            );
            if (team.trim().toLowerCase() === 'back') {
              console.log(chalk.yellow('Cancelled. Back to selector selection.'));
              console.log('\n');
              break;
            }
            if (!team.trim()) {
              console.log(chalk.red('Please enter a team.'));
              continue;
            }
            console.log(chalk.blue(`Team:`), `${chalk.green(`${chalk.bold(team)}`)}`);
            console.log('\n');
            addedSelectors.push(`team=${team}`);
            break;
          }
          break;
        }
        case 'limit': {
          while (true) {
            const limit = await createQuestion(chalk.cyan('Limit(int). Type "back" to go back: '));
            if (limit.trim().toLowerCase() === 'back') {
              console.log(chalk.yellow('Cancelled. Back to selector selection.'));
              console.log('\n');
              break;
            }
            if (!limit.trim()) {
              console.log(chalk.red('Please enter a limit.'));
              continue;
            }
            console.log(chalk.blue(`Limit:`), `${chalk.green(`${chalk.bold(limit)}`)}`);
            console.log('\n');
            addedSelectors.push(`limit=${limit}`);
            break;
          }
          break;
        }
        case 'level': {
          while (true) {
            const level = await createQuestion(
              chalk.cyan(
                "Exp Level(int or range format: '10' or '10..20'). Type \"back\" to go back: "
              )
            );
            if (level.trim().toLowerCase() === 'back') {
              console.log(chalk.yellow('Cancelled. Back to selector selection.'));
              console.log('\n');
              break;
            }
            if (!level.trim()) {
              console.log(chalk.red('Please enter a level.'));
              continue;
            }
            console.log(chalk.blue(`Level:`), `${chalk.green(`${chalk.bold(level)}`)}`);
            console.log('\n');
            addedSelectors.push(`level=${level}`);
            break;
          }
          break;
        }
        case 'gamemode': {
          const gamemodeList = [
            'survival',
            'creative',
            'adventure',
            'spectator',
            'back - Go back to selector selection',
          ];
          const gamemodeResult = await selectFromList(
            chalk.cyan('Player Gamemode: '),
            gamemodeList
          );
          const gamemode = gamemodeResult.split(' ')[0];

          if (gamemode.toLowerCase() === 'back') {
            console.log(chalk.yellow('Cancelled. Back to selector selection.'));
            console.log('\n');
            break;
          }

          console.log(chalk.blue(`Gamemode:`), `${chalk.green(`${chalk.bold(gamemode)}`)}`);
          console.log('\n');
          addedSelectors.push(`gamemode=${gamemode}`);
          break;
        }
        case 'advancements': {
          while (true) {
            const advancement = await createQuestion(
              chalk.cyan(
                'Advancement(format: <advancement_ID>=true/false). Type "back" to go back: '
              )
            );
            if (advancement.trim().toLowerCase() === 'back') {
              console.log(chalk.yellow('Cancelled. Back to selector selection.'));
              console.log('\n');
              break;
            }
            if (!advancement.trim()) {
              console.log(chalk.red('Please enter an advancement.'));
              continue;
            }

            console.log(
              chalk.blue(`Advancements:`),
              `${chalk.green(`${chalk.bold(advancement)}`)}`
            );
            console.log('\n');
            addedSelectors.push(`advancements=${advancement}`);
            break;
          }
          break;
        }
        case 'predicate': {
          while (true) {
            const predicate = await createQuestion(
              chalk.cyan(
                'Predicate(predicate_id | If it put \'!\' at the beginning, the predicate will be excluded.). Type "back" to go back: '
              )
            );
            if (predicate.trim().toLowerCase() === 'back') {
              console.log(chalk.yellow('Cancelled. Back to selector selection.'));
              console.log('\n');
              break;
            }
            if (!predicate.trim()) {
              console.log(chalk.red('Please enter a predicate.'));
              continue;
            }

            console.log(chalk.blue(`Predicate:`), `${chalk.green(`${chalk.bold(predicate)}`)}`);
            console.log('\n');
            addedSelectors.push(`predicate=${predicate}`);
            break;
          }
          break;
        }
        case 'sort': {
          const sortList = [
            'nearest - Select from the nearest entity first',
            'furthest - Select from the furthest entity first',
            'random - Select random',
            'arbitrary - Select by spawn time',
            'back - Go back to selector selection',
          ];
          const sortResult = await selectFromList(chalk.cyan('Sort: '), sortList);
          const sort = sortResult.split(' ')[0];

          if (sort.toLowerCase() === 'back') {
            console.log(chalk.yellow('Cancelled. Back to selector selection.'));
            console.log('\n');
            break;
          }

          console.log(chalk.blue(`Sort:`), `${chalk.green(`${chalk.bold(sort)}`)}`);
          console.log('\n');
          addedSelectors.push(`sort=${sort}`);
          break;
        }
        case 'OK': {
          continueAdding = false;
          break;
        }
      }
    }
    addSelectors = addedSelectors.join(',');

    console.log(
      chalk.blue(`All selectors:`),
      `${chalk.green(`${chalk.bold(addSelectors || null)}`)}`
    );
    console.log('\n');
  } else {
    console.log(
      `${chalk.blue('Additional target selectors:')} ${chalk.green(`${chalk.bold('No')}`)}`
    );
    console.log('\n');
  }

  const addedSelectorsTF: boolean = !!addSelectors;

  if (addedSelectorsTF) {
    selector = `${targetType}[${addSelectors}]`;
  } else {
    selector = `${targetType}`;
  }

  return selector;
}

File: cli/cmd/selector.go
package cmd

import (
	"fmt"
	"strings"

	"github.com/charmbracelet/bubbles/textinput"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/huh"
	"github.com/fatih/color"
)

type CustomSelectorVal struct {
	Key   string
	Value string
}

func addSelectorsQuestion() (string, error) {
	for {
		targetOptions := []string{
			"@p - Near Player",
			"@a - All Player",
			"@s - Myself",
			"@r - Random Player",
			"@n - Nearest Player (1.21+)",
			"PlayerName",
		}

		var choice string
		err := huh.NewSelect[string]().
			Title("Select a target selector type:").
			Options(huh.NewOptions(targetOptions...)...).
			Value(&choice).
			Run()
		if err != nil {
			return "", err
		}

		targetType := strings.Split(choice, " ")[0]

		if targetType == "PlayerName" {
			var name string
			err = huh.NewInput().
				Title("Enter the player name (Type \"back\" to go back).\nNote: Use '!' prefix (e.g. !notch) to skip Mojang verification:").
				Value(&name).
				Validate(func(s string) error {
					if strings.TrimSpace(s) == "" {
						return fmt.Errorf("player name cannot be empty")
					}
					return nil
				}).
				Run()
			if err != nil {
				return "", err
			}

			name = strings.TrimSpace(name)
			if strings.ToLower(name) == "back" {
				continue
			}

			if strings.HasPrefix(name, "!") {
				targetType = strings.TrimPrefix(name, "!")
			} else {
				fmt.Println(color.CyanString("Checking player existence..."))
				if mojangPlayerExists(name) {
					fmt.Println(color.GreenString("Player %s exists in Mojang database!", name))
					targetType = name
				} else {
					fmt.Println(color.YellowString("WARN: Player %s not found in Mojang database.", name))
					var useAnyway bool
					err = huh.NewConfirm().
						Title("Use this name anyway?").
						Value(&useAnyway).
						Run()
					if err != nil {
						return "", err
					}
					if useAnyway {
						targetType = name
					} else {
						continue
					}
				}
			}
		}

		greenBold := color.New(color.FgGreen, color.Bold)
		fmt.Println(color.BlueString("Target:"), greenBold.Sprint(targetType))

		if !strings.HasPrefix(targetType, "@") {
			fmt.Println(color.BlueString("Additional target selectors:"), color.GreenString("Skipped (Specific Player Name)"))
			return targetType, nil
		}

		var refine bool
		err = huh.NewConfirm().
			Title("Do you want to add more target selectors? (distance, score, tag, team, limit, etc.)").
			Value(&refine).
			Run()
		if err != nil {
			return "", err
		}

		if !refine {
			return targetType, nil
		}

		allSelectorTypes := []string{
			"distance",
			"score",
			"tag",
			"team",
			"limit",
			"level",
			"gamemode",
			"advancements",
			"predicate",
			"sort",
		}

		descriptions := map[string]string{
			"distance":     "Distance to Entity(=Player)",
			"score":        "The score value or range",
			"tag":          "The tag which the entity has",
			"team":         "The team which the entity joins",
			"limit":        "Amount limit",
			"level":        "Experience level",
			"gamemode":     "Player gamemode",
			"advancements": "The advancements which the player has",
			"predicate":    "Match predicates",
			"sort":         "Specify selection sorting order",
		}

		var configuredCustom []CustomSelectorVal

		for {
			var available []string
			for _, s := range allSelectorTypes {
				alreadyAdded := false
				for _, added := range configuredCustom {
					if added.Key == s {
						alreadyAdded = true
						break
					}
				}
				if !alreadyAdded {
					available = append(available, s)
				}
			}

			var menuChoices []huh.Option[string]

			// 1st item: Added selectors
			addedLabel := fmt.Sprintf("[Added Selectors] (%d configured)", len(configuredCustom))
			menuChoices = append(menuChoices, huh.NewOption(addedLabel, "__ADDED__"))

			// 2nd item onwards: Unconfigured custom selectors
			for _, av := range available {
				menuChoices = append(menuChoices, huh.NewOption(fmt.Sprintf("%s - %s", av, descriptions[av]), av))
			}

			// Last item: OK
			menuChoices = append(menuChoices, huh.NewOption("[OK - Finish]", "__OK__"))

			var choice string
			err = huh.NewSelect[string]().
				Title("Selector Configuration Menu:").
				Options(menuChoices...).
				Value(&choice).
				Run()
			if err != nil {
				return "", err
			}

			if choice == "__OK__" {
				break
			}

			if choice == "__ADDED__" {
				if len(configuredCustom) == 0 {
					fmt.Println(color.YellowString("No custom selectors configured yet."))
					continue
				}

				ti := textinput.New()
				ti.Placeholder = "New value"
				ti.CharLimit = 156
				ti.Width = 40

				m := addedSelectorsModel{
					options:   configuredCustom,
					textInput: ti,
				}

				p := tea.NewProgram(m, tea.WithAltScreen())
				res, err := p.Run()
				if err != nil {
					return "", err
				}

				finalModel := res.(addedSelectorsModel)
				configuredCustom = finalModel.options
				continue
			}

			// Add new custom selector
			selKey := choice
			var val string
			switch selKey {
			case "distance":
				err = huh.NewInput().Title("Distance (e.g. 1..5):").Value(&val).Run()
			case "score":
				err = huh.NewInput().Title("Score format (e.g. score=1..10):").Value(&val).Run()
			case "tag":
				err = huh.NewInput().Title("Tag (e.g. tag_name or !tag_name):").Value(&val).Run()
			case "team":
				err = huh.NewInput().Title("Team (e.g. team_name or !team_name):").Value(&val).Run()
			case "limit":
				err = huh.NewInput().Title("Limit (e.g. 1):").Value(&val).Run()
			case "level":
				err = huh.NewInput().Title("Exp level (e.g. 10 or 10..20):").Value(&val).Run()
			case "gamemode":
				err = huh.NewSelect[string]().
					Title("Player gamemode:").
					Options(huh.NewOptions("survival", "creative", "adventure", "spectator")...).
					Value(&val).
					Run()
			case "advancements":
				err = huh.NewInput().Title("Advancement (e.g. adv_id=true):").Value(&val).Run()
			case "predicate":
				err = huh.NewInput().Title("Predicate (e.g. predicate_id):").Value(&val).Run()
			case "sort":
				err = huh.NewSelect[string]().
					Title("Sort order:").
					Options(huh.NewOptions("nearest", "furthest", "random", "arbitrary")...).
					Value(&val).
					Run()
			}
			if err != nil {
				return "", err
			}

			val = strings.TrimSpace(val)
			if val != "" {
				configuredCustom = append(configuredCustom, CustomSelectorVal{
					Key:   selKey,
					Value: val,
				})
				fmt.Println(color.GreenString("Added %s: %s", selKey, val))
			}
		}

		if len(configuredCustom) > 0 {
			var parts []string
			for _, sc := range configuredCustom {
				parts = append(parts, fmt.Sprintf("%s=%s", sc.Key, sc.Value))
			}
			targetType = fmt.Sprintf("%s[%s]", targetType, strings.Join(parts, ","))
		}
		return targetType, nil
	}
}

// bubbletea model for added custom target selector parameters
type addedSelectorsModel struct {
	options   []CustomSelectorVal
	cursor    int
	textInput textinput.Model
	editing   bool
}

func (m addedSelectorsModel) Init() tea.Cmd {
	return nil
}

func (m addedSelectorsModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	var cmd tea.Cmd

	if m.editing {
		switch msg := msg.(type) {
		case tea.KeyMsg:
			switch msg.String() {
			case "enter":
				val := strings.TrimSpace(m.textInput.Value())
				if val != "" {
					m.options[m.cursor].Value = val
				}
				m.editing = false
				m.textInput.Blur()
				return m, nil

			case "esc":
				m.editing = false
				m.textInput.Blur()
				return m, nil
			}
		}
		m.textInput, cmd = m.textInput.Update(msg)
		return m, cmd
	}

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.String() {
		case "ctrl+c", "q", "esc", "enter":
			return m, tea.Quit

		case "up", "k":
			if m.cursor > 0 {
				m.cursor--
			}

		case "down", "j":
			if m.cursor < len(m.options)-1 {
				m.cursor++
			}

		case "e":
			if len(m.options) > 0 {
				m.editing = true
				m.textInput.SetValue(m.options[m.cursor].Value)
				m.textInput.Focus()
				return m, textinput.Blink
			}

		case "d":
			if len(m.options) > 0 {
				m.options = append(m.options[:m.cursor], m.options[m.cursor+1:]...)
				if m.cursor >= len(m.options) && m.cursor > 0 {
					m.cursor = len(m.options) - 1
				}
				if len(m.options) == 0 {
					return m, tea.Quit
				}
			}
		}
	}
	return m, nil
}

func (m addedSelectorsModel) View() string {
	var s strings.Builder
	fmt.Fprintf(&s, "\n  %s\n\n", color.CyanString("Configure Added Selectors:"))

	if len(m.options) == 0 {
		s.WriteString("    No custom selectors configured yet.\n")
	} else {
		for i, sc := range m.options {
			if m.cursor == i {
				s.WriteString(fmt.Sprintf("  %s %s=%s\n",
					color.CyanString(">"),
					color.CyanString(sc.Key),
					color.YellowString(sc.Value)))
			} else {
				s.WriteString(fmt.Sprintf("    %s=%s\n", sc.Key, sc.Value))
			}
		}
	}

	if m.editing {
		fmt.Fprintf(&s, "\n  %s\n", color.GreenString("Editing value for %s:", m.options[m.cursor].Key))
		fmt.Fprintf(&s, "  %s\n", m.textInput.View())
		fmt.Fprintf(&s, "\n  %s\n", color.HiBlackString("(press 'enter' to save, 'esc' to cancel)"))
	} else {
		fmt.Fprintf(&s, "\n  %s\n", color.HiBlackString("(press 'e' to edit, 'd' to delete, 'q' or 'enter' to go back)"))
	}

	return s.String()
}
